### PR TITLE
APIv2 restplus free

### PR DESCRIPTION
**Pull request by quaxsze**

### CODE REVIEW

**Review by maudetes — COMMENTED** [review]
A very first batch of review

**Review by maudetes — COMMENTED** [review]
We have an issue related on a hack to use a custom page size on search v2 endpoints.
It now fails with the new apiv2, could you take a look? https://github.com/etalab/data.gouv.fr/issues/880

File: udata/api/fields.py [hunk 1]
```diff
@@ -90,3 +91,71 @@ def pager(page_fields):
             description='The previous page URL if exists')
     }
     return pager_fields
+
+
+class MaImageField(ma_fields.Field):
```
Comment (maudetes) [inline udata/api/fields.py:96]: Maybe we want to change the `Ma*Field` pattern, but don't have perfect ideas. How about `Marsh*Field`

File: udata/api/specs.py [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+from udata.core.dataset.apiv2 import get_dataset_search, get_resources_paginated, get_specific_dataset_by_id, get_specific_resource_by_rid
+from udata.core.organization.apiv2 import get_organization_search
+from udata.core.reuse.apiv2 import get_reuse_search
+from udata.core.user.apiv2 import get_users_list, post_new_user
+
+
+def init_app(app):
+    app.config.update({
+        'APISPEC_SPEC': APISpec(
+            title="Udata",
```
Comment (maudetes) [inline udata/api/specs.py:15]:
```suggestion
            title="udata",
```

File: udata/api/specs.py [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+            version="1.0.0",
+            openapi_version="3.0.2",
+            plugins=[MarshmallowPlugin()],
+        ),
+        'APISPEC_SWAGGER_URL': '/swagger/',
+        'APISPEC_SWAGGER_UI_URL': '/swagger-ui/',
```
Comment (maudetes) [inline udata/api/specs.py:21]: I think we don't need a swagger ui, if I understand correctly, could we deactivate it?

File: udata/api/specs.py [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+
+def init_app(app):
+    app.config.update({
+        'APISPEC_SPEC': APISpec(
+            title="Udata",
+            version="1.0.0",
+            openapi_version="3.0.2",
+            plugins=[MarshmallowPlugin()],
+        ),
+        'APISPEC_SWAGGER_URL': '/swagger/',
```
Comment (maudetes) [inline udata/api/specs.py:20]: Should we set it to be on `api/2` instead?

File: udata/api/specs.py [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+        'APISPEC_SWAGGER_UI_URL': '/swagger-ui/',
+    })
+
+    docs = FlaskApiSpec(app)
+    with app.app_context():
+        docs.register(get_dataset_search, endpoint='apiv2.dataset_search')
```
Comment (maudetes) [inline udata/api/specs.py:26]: I'm wondering if we could use a decorator instead, so that we don't forget to add this doc register?

File: udata/core/user/apiv2_schemas.py [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+    id = fields.Str(dump_only=True)
+    first_name = fields.Str(dump_only=True)
+    last_name = fields.String(dump_only=True)
+    slug = fields.String(required=True)
+    about = fields.String()
+    page = MaURLFor(endpoint='users.show', mapper=lambda u: {'user': u}, fallback_endpoint='api.user', dump_only=True)
+    uri = MaURLFor(endpoint='api.user', mapper=lambda u: {'user': u}, dump_only=True)
+    avatar = MaImageField(dump_only=True)
+    avatar_thumbnail = MaImageField(dump_only=True, attribute='avatar', size=BIGGEST_AVATAR_SIZE)
+    email = fields.Function(lambda o: o.email if current_user_is_admin_or_self() else None, dump_only=True),
```
Comment (maudetes) [inline udata/core/user/apiv2_schemas.py:32]: email is not showing somehow

File: udata/api/specs.py (hunk 4 above)
Comment (maudetes) [reply udata/api/specs.py:20]: Actual swagger url is `/api/1/swagger.json` for api v1

File: udata/api/specs.py (hunk 3 above)
Comment (maudetes) [reply udata/api/specs.py]: We actually could have one to access it locally or on dev env.
I set it back, but it showed "No API definition provided.". I tried to look into it, but did not succeed. I think we can do without it for now but should keep it in mind.

**Review by maudetes — COMMENTED** [review]
Nice! I find the code readable and it seems to work nicely. :clap: 

I have some comments, but mostly non blocking ones (except for the one on `secure`).

I wish we could have more tests on these endpoints, could we duplicate some tests that exists for api v1 maybe?

File: udata/api/fields.py [hunk 7]
```diff
@@ -90,3 +91,67 @@ def pager(page_fields):
...
+            return
+        if self.size:
+            return value(self.size, external=True)
+        else:
+            return value.fs.url(value.filename, external=True)
+
+
+class MarshURLFor(ma_fields.Field):
+
+    _CHECK_ATTRIBUTE = False
```
[the remark this answers is not in the document]
Comment (quaxsze) [reply udata/api/fields.py:113]: I quote the doc:
```
Some fields, such as Method fields and Function fields, are not expected
to exist as attributes on the objects to serialize. Set this to False
for those fields
```

File: udata/api/__init__.py [hunk 8]
```diff
@@ -302,14 +259,162 @@ def marshal_page_with(func):
...
+@apiv2_blueprint.route('/', endpoint='root')
+def apiv2_root():
+    '''
+    Useful to return the api v2 root url
+    '''
+    abort(404)
+
+
+@apiv2_blueprint.before_request
+def authentifyV2():
```
[the remark this answers is not in the document]
Comment (quaxsze) [reply udata/api/__init__.py:342]: Before every request, as it is for APIv1

File: udata/core/user/apiv2.py [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+        else:
+            return search_users.order_by('$text_score').paginate(kwargs['page'], kwargs['page_size'])
+    if kwargs.get('sort'):
+        return users.order_by(kwargs['sort']).paginate(kwargs['page'], kwargs['page_size'])
+    return users.order_by(DEFAULT_SORTING).paginate(kwargs['page'], kwargs['page_size'])
+
+
+@apiv2.route('/users/', endpoint='post_user', methods=['POST'])
+@csrf.exempt
+@UDataApiV2.secure
```
[the remark this answers is not in the document]
Comment (maudetes) [reply udata/core/user/apiv2.py:53]: My bad, it was my request app that remembered cookies of previous requests :p
It correctly returns a 401 :)

File: udata/tests/apiv2/test_datasets.py [hunk 10]
```diff
@@ -81,7 +81,7 @@ def test_get_missing_param(self):
 
     def test_get_next_page(self):
         '''Should fetch 2 pages of resources from the API'''
-        resources = [ResourceFactory() for _ in range(80)]
+        resources = [ResourceFactory() for _ in range(40)]
```
[the remark this answers is not in the document]
Comment (abulte) [reply udata/tests/apiv2/test_datasets.py:84]: 👍 probably because of https://github.com/opendatateam/udata/pull/2755/files#diff-a62c9bd3310470c9b7bfe01b30f76621a938f5d5be834d2b479dafd146b3d154R16 ?

File: udata/tests/apiv2/test_datasets.py [hunk 11]
```diff
@@ -96,7 +96,6 @@ def test_get_next_page(self):
         response = self.get(data['next_page'])
         self.assert200(response)
         data = response.json
-        assert len(data['data']) == len(resources) - DEFAULT_PAGE_SIZE
```
Comment (abulte) [inline udata/tests/apiv2/test_datasets.py:99]: This should still pass?

**Review by abulte — COMMENTED** [review]
This is cool!

We need some tests on the new users v2 API I think
